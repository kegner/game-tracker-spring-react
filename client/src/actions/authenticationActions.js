export const SET_USER = "SET_USER";
export const CLEAR_USER = "CLEAR_USER";

export const setUserData = (user) => ({
  type: SET_USER,
  payload: { user }
});

export const clearUserData = () => ({
    type: CLEAR_USER
});