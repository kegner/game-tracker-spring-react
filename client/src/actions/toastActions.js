export const SHOW_SUCCESS = "SHOW_SUCCESS";
export const SHOW_ERROR = "SHOW_ERROR";
export const HIDE_TOAST = "HIDE_TOAST";

export const showSuccessToast = (message) => ({
  type: SHOW_SUCCESS,
  payload: { message }
});

export const showErrorToast = (message) => ({
  type: SHOW_ERROR,
  payload: { message }
});

export const hideToast = (id) => ({
  type: HIDE_TOAST,
  payload: { id }
});