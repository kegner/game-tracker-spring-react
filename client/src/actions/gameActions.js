import axios from 'axios';
import { showSuccessToast, showErrorToast } from './toastActions';

export const SELECT_GAME = "SELECT_GAME";
export const FETCH_GAMES = "FETCH_GAMES";
export const CREATE_GAME = "CREATE_GAME";
export const UPDATE_GAME = "UPDATE_GAME";
export const DELETE_GAME = "DELETE_GAME";
export const START_LOADING = "START_LOADING";
export const START_SAVING = "START_SAVING";
export const STOP_LOADING = "STOP_LOADING";
export const STOP_SAVING = "STOP_SAVING";

const fetchSuccess = (games) => ({
  type: FETCH_GAMES,
  payload: { games }
});

const createSuccess = (game) => ({
  type: CREATE_GAME,
  payload: { game }
});

const updateSuccess = (game) => ({
  type: UPDATE_GAME,
  payload: { game }
});

const deleteSuccess = (id) => ({
  type: DELETE_GAME,
  payload: { id }
});

const startLoading = () => ({
  type: START_LOADING
});

const startSaving = () => ({
  type: START_SAVING
});

const stopLoading = () => ({
  type: STOP_LOADING
});

const stopSaving = () => ({
  type: STOP_SAVING
});

const errorHandler = (error, dispatch) => {
  const validResponses = [404, 409];
  if (error && error.response && validResponses.includes(error.response.status)) {
    // this is the expected error format for Spring exception messages
    dispatch(showErrorToast(error.response.data.message));
  } else {
    console.error(error);
  }
};

export const selectGame = (game) => ({
  type: SELECT_GAME,
  payload: { game }
});

export const fetchGames = (userId) => async (dispatch) => {
  dispatch(startLoading());

  try {
    const res = await axios.get("/api/v1/games", { params: { userId } });
    dispatch(fetchSuccess(res.data.games));
  } catch (error) {
    dispatch(stopLoading());
    errorHandler(error, dispatch);
  }
};

export const createGame = (game) => async (dispatch) => {
  dispatch(startSaving());

  try {
    const res = await axios.post("/api/v1/games", game);

    dispatch(createSuccess(res.data));
    dispatch(showSuccessToast("Game added."));
  } catch (error) {
    dispatch(stopSaving());
    errorHandler(error, dispatch);
  }
};

export const updateGame = (game) => async (dispatch) => {
  dispatch(startSaving());

  try {
    const res = await axios.put("/api/v1/games", game);

    dispatch(updateSuccess(res.data));
    dispatch(showSuccessToast("Game updated."));
  } catch (error) {
    dispatch(stopSaving());
    errorHandler(error, dispatch);
  }
};

export const deleteGame = (id) => async (dispatch) => {
  dispatch(startSaving());

  try {
    await axios.delete(`/api/v1/games/${id}`);
    dispatch(deleteSuccess(id));
    dispatch(showSuccessToast("Game deleted."));
  } catch (error) {
    dispatch(stopSaving());
    errorHandler(error, dispatch);
  }
};