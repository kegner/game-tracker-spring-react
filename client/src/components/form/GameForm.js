import React, { useEffect, useState } from 'react'
import { createGame, updateGame } from '../../actions/gameActions';
import { useDispatch, useSelector } from 'react-redux';
import FormItem from './FormItem';
import Spinner from '../layout/Spinner';

const GameForm = () => {

  const [id, setId] = useState("");
  const [title, setTitle] = useState("");
  const [platform, setPlatform] = useState("");
  const [startDate, setStartDate] = useState("");
  const [endDate, setEndDate] = useState("");
  const [status, setStatus] = useState("");
  const [hoursToComplete, setHoursToComplete] = useState(0);
  const [rating, setRating] = useState(0);
  const [price, setPrice] = useState(0);

  const { selectedGame, saving } = useSelector(state => state.gameReducer);
  const { user } = useSelector(state => state.authenticationReducer);

  const dispatch = useDispatch();

  useEffect(() => {
    setId(selectedGame.id || "");
    setTitle(selectedGame.title || "");
    setPlatform(selectedGame.platform || "");
    setStartDate(selectedGame.startDate || "");
    setEndDate(selectedGame.endDate || "");
    setStatus(selectedGame.status || "");
    setHoursToComplete(selectedGame.hoursToComplete || 0);
    setRating(selectedGame.rating || 0);
    setPrice(selectedGame.price || 0);
  }, [selectedGame]);

  const createNew = (e) => {
    e.preventDefault();
    const game = {
      title,
      platform,
      startDate,
      endDate,
      status,
      hoursToComplete,
      rating,
      price,
      userId: user.id
    };
    dispatch(createGame(game));
  }

  const update = (e) => {
    e.preventDefault();
    const game = {
      id,
      title,
      platform,
      startDate,
      endDate,
      status,
      hoursToComplete,
      rating,
      price,
      userId: user.id
    };
    dispatch(updateGame(game));
  }

  const setState = (e, setter) => {
    setter(e.target.value);
  }

  const spinner = saving ? <Spinner /> : <></>;

  return (
    <div className="position-relative">
      <div className="card">
        <form>
          <div className="card-header bg-dark text-white">
            Game Form
          </div>
          <div className="card-body">
            <input type="hidden" value={id} />
            <div className="row mb-2">
              <div className="col-3">
                <FormItem id="title" label="Title" value={title} type="text"
                  onChange={(e) => setState(e, setTitle)} />
              </div>
              <div className="col-3">
                <FormItem id="platform" label="Platform" value={platform} type="text"
                  onChange={(e) => setState(e, setPlatform)} />
              </div>
              <div className="col-3">
                <FormItem id="startDate" label="Start Date" value={startDate} type="date"
                  onChange={(e) => setState(e, setStartDate)} />
              </div>
              <div className="col-3">
                <FormItem id="endDate" label="End Date" value={endDate} type="date"
                  onChange={(e) => setState(e, setEndDate)} />
              </div>
            </div>
            <div className="row mb-2">
              <div className="col-3">
                <FormItem id="status" label="Status" value={status} type="text"
                  onChange={(e) => setState(e, setStatus)} />
              </div>
              <div className="col-3">
                <FormItem id="hoursToComplete" label="Hours to Complete" value={hoursToComplete} type="number"
                  onChange={(e) => setState(e, setHoursToComplete)} />
              </div>
              <div className="col-3">
                <FormItem id="rating" label="Rating" value={rating} type="number"
                  onChange={(e) => setState(e, setRating)} />
              </div>
              <div className="col-3">
                <FormItem id="price" label="Price" value={price} type="number" currency={true}
                  onChange={(e) => setState(e, setPrice)} />
              </div>
            </div>
          </div>
          <div className="card-footer bg-dark text-white">
            <div className="d-flex justify-content-end align-items-end">
              <button type="submit" className="btn btn-primary" disabled={title === ""} onClick={createNew}>
                Create New
              </button>
              <div className="mx-2"></div>
              <button type="submit" className="btn btn-primary" disabled={id === ""} onClick={update}>
                Update
              </button>
            </div>
          </div>
        </form>
      </div>
      {spinner}
    </div>
  )
}

export default GameForm
