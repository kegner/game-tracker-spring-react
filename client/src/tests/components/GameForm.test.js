import React from 'react';
import { Provider } from "react-redux";
import { render, getByText, fireEvent } from '@testing-library/react';
import { START_SAVING, CREATE_GAME, UPDATE_GAME } from '../../actions/gameActions';
import { SHOW_SUCCESS } from '../../actions/toastActions';
import { createTestStore } from '../store/testStore';
import axios from 'axios';
import GameForm from '../../components/form/GameForm';

let store;
let dispatch;
let component;
let container;
let asFragment;
let game = {
  id: "123", title: "My Game"
};

jest.mock("axios");

describe("GameForm", () => {

  beforeEach(() => {
    store = createTestStore({
      mockStore: true,
      initialState: {
        gameReducer: {
          selectedGame: game
        },
        authenticationReducer: {
          user: {
            id: 1
          }
        }
      }
    });
    dispatch = store.dispatch;
    component = render(
      <Provider store={store}>
        <GameForm />
      </Provider>
    );
    container = component.container;
    asFragment = component.asFragment;
  });

  it("exists", () => {
    expect(asFragment()).toMatchSnapshot();
  });

  it("create new game", async () => {
    axios.post.mockResolvedValue({ data: game });
    const expectedActions = [
      { type: START_SAVING },
      { type: CREATE_GAME, payload: { game } },
      { type: SHOW_SUCCESS, payload: { message: "Game added." } }
    ];

    const button = getByText(container, "Create New");
    expect(dispatch).not.toHaveBeenCalled();
    fireEvent.click(button);
    // await for all dispatches
    await new Promise(r => setTimeout(r, 1));
    expect(dispatch).toHaveBeenCalledTimes(1);
    expect(store.getActions()).toEqual(expectedActions);
  });

  it("update game", async () => {
    axios.put.mockResolvedValue({ data: game });
    const expectedActions = [
      { type: START_SAVING },
      { type: UPDATE_GAME, payload: { game } },
      { type: SHOW_SUCCESS, payload: { message: "Game updated." } }
    ];

    const button = getByText(container, "Update");
    expect(dispatch).not.toHaveBeenCalled();
    fireEvent.click(button);
    // await for all dispatches
    await new Promise(r => setTimeout(r, 1));
    expect(dispatch).toHaveBeenCalledTimes(1);
    expect(store.getActions()).toEqual(expectedActions);
  });

});