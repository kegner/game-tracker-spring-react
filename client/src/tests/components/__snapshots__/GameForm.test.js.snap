// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameForm exists 1`] = `
<DocumentFragment>
  <div
    class="position-relative"
  >
    <div
      class="card"
    >
      <form>
        <div
          class="card-header bg-dark text-white"
        >
          Game Form
        </div>
        <div
          class="card-body"
        >
          <input
            type="hidden"
            value="123"
          />
          <div
            class="row mb-2"
          >
            <div
              class="col-3"
            >
              <label
                class="form-label"
                for="title-1"
              >
                Title
              </label>
              <input
                class="form-control undefined"
                id="title-1"
                type="text"
                value="My Game"
              />
            </div>
            <div
              class="col-3"
            >
              <label
                class="form-label"
                for="platform-2"
              >
                Platform
              </label>
              <input
                class="form-control undefined"
                id="platform-2"
                type="text"
                value=""
              />
            </div>
            <div
              class="col-3"
            >
              <label
                class="form-label"
                for="startDate-3"
              >
                Start Date
              </label>
              <input
                class="form-control undefined"
                id="startDate-3"
                type="date"
                value=""
              />
            </div>
            <div
              class="col-3"
            >
              <label
                class="form-label"
                for="endDate-4"
              >
                End Date
              </label>
              <input
                class="form-control undefined"
                id="endDate-4"
                type="date"
                value=""
              />
            </div>
          </div>
          <div
            class="row mb-2"
          >
            <div
              class="col-3"
            >
              <label
                class="form-label"
                for="status-5"
              >
                Status
              </label>
              <input
                class="form-control undefined"
                id="status-5"
                type="text"
                value=""
              />
            </div>
            <div
              class="col-3"
            >
              <label
                class="form-label"
                for="hoursToComplete-6"
              >
                Hours to Complete
              </label>
              <input
                class="form-control undefined"
                id="hoursToComplete-6"
                type="number"
                value="0"
              />
            </div>
            <div
              class="col-3"
            >
              <label
                class="form-label"
                for="rating-7"
              >
                Rating
              </label>
              <input
                class="form-control undefined"
                id="rating-7"
                type="number"
                value="0"
              />
            </div>
            <div
              class="col-3"
            >
              <label
                class="form-label"
                for="price-8"
              >
                Price
              </label>
              <input
                class="form-control undefined"
                id="price-8"
                type="number"
                value="0"
              />
            </div>
          </div>
        </div>
        <div
          class="card-footer bg-dark text-white"
        >
          <div
            class="d-flex justify-content-end align-items-end"
          >
            <button
              class="btn btn-primary"
              type="submit"
            >
              Create New
            </button>
            <div
              class="mx-2"
            />
            <button
              class="btn btn-primary"
              type="submit"
            >
              Update
            </button>
          </div>
        </div>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
